### PR TITLE
sof performance build enabling

### PR DESCRIPTION
**Pull request by btian1**
Enable a specific performance profiling build to cover core/task/component
performance, this build use option -o app/perf_overlay.conf to enable.
it utilize perf_cnt module to measure performance.

Signed-off-by: btian1 <baofeng.tian@intel.com>

### CODE REVIEW

File: scripts/xtensa-build-zephyr.py [hunk 1]
```diff
@@ -185,6 +185,8 @@ def parse_args():
 						default=[], help="Paths to overlays")
 	parser.add_argument("-p", "--pristine", required=False, action="store_true",
 						help="Perform pristine build removing build directory.")
+	parser.add_argument("-s", "--perf", required=False, action="store_true",
+						help="enable performance build for core/task/module.")
```
Comment (marc-hb) [inline scripts/xtensa-build-zephyr.py:189]: There are only 26 letters and I don't think we should spend one for each overlay. It barely saves any typing compared to `-o app/perf<TAB>` and it's an overlay that few people (if any) will use regularly.

**Review by marc-hb — COMMENTED** [review]
Not sure how many independent changes there  here but for sure this requires more than a single commit https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes

EDIT: always start a PR as a Github draft. This avoids reviewers seeing so many build failures.

File: scripts/xtensa-build-zephyr.py (hunk 1 above)
Comment (btian1) [reply scripts/xtensa-build-zephyr.py:189]: performance build is important, and for convinence, let's first use -s first, current only 7 chars are used, in future, if there is a  
necessary, let's combine or use -o, for now, let's keep it, since this is the first time for enabling.

**Comment by btian1** [comment]
> 

Thanks, Fixed

**Comment by btian1** [comment]
> 

Sorry, first time uploading to sof， there is some build errors, i will fix them EOW

File: scripts/xtensa-build-zephyr.py (hunk 1 above)
Comment (marc-hb) [reply scripts/xtensa-build-zephyr.py:189]: For the moment it's important... to you. Once enough people complain that typing `-o app/perf<TAB>` takes too much of their time then we can discuss again. This seems unlikely to me; I expect most people to use the "up" arrow most of the time.

File: xtos/include/sof/lib/perf_cnt.h [hunk 2]
```diff
@@ -111,16 +132,12 @@ struct perf_cnt_data {
...
-			(pcd)->cpu_delta_last = cpu_ts - (pcd)->cpu_ts;		\
-		}								\
-		(pcd)->plat_ts = plat_ts;					\
-		(pcd)->cpu_ts = cpu_ts;						\
+		uint64_t plat_ts =						\
+			(uint64_t)sof_cycle_get_64();				\
+		uint64_t cpu_ts =						\
+			(uint64_t)perf_cnt_get_cpu_ts();			\
+		(pcd)->plat_delta_last = (uint32_t)(plat_ts - (pcd)->plat_ts);	\
+		(pcd)->cpu_delta_last = (uint32_t)(cpu_ts - (pcd)->cpu_ts);		\
```
Comment (marc-hb) [inline xtos/include/sof/lib/perf_cnt.h:140]: No whitespace changes, they make it much harder to see what actually changed.

File: xtos/include/sof/lib/perf_cnt.h [hunk 3]
```diff
@@ -17,8 +17,8 @@
 #include <rtos/timer.h>
 
 struct perf_cnt_data {
-	uint32_t plat_ts;
-	uint32_t cpu_ts;
+	uint64_t plat_ts;
+	uint64_t cpu_ts;
```
Comment (marc-hb) [inline xtos/include/sof/lib/perf_cnt.h:21]: You must explain real code changes like this one.

The commit message sounds like it's merely adding a new build configuration but this does more than that.

This most likely requires multiple commits https://wiki.openstack.org/wiki/GitCommitMessages#Structural_split_of_changes

**Review by marc-hb — COMMENTED** [review]
A "performance build" sounds like a faster build. You mean a "Performance analysis" build or a "performance profiling" build, not a "performance" build.

Here are some terminology examples:
https://en.wikipedia.org/wiki/Perf_(Linux)
https://www.intel.com/content/www/us/en/develop/documentation/vtune-help/top/set-up-analysis-target/linux-targets/compiler-switches-for-perf-analysis-linux-targets.html

**Comment by btian1** [comment]
Please review, the error is not introduced by this PR, it is already there before the rule(checkpatch_list.sh) created.

File: app/perf_overlay.conf [hunk 4]
```diff
@@ -0,0 +1,2 @@
+CONFIG_PERFORMANCE_COUNTERS=y
```
Comment (ranj063) [inline app/perf_overlay.conf:1]: @btian1 Can you split this patch into 2 separate patches with the first one to introduce the macros and the second one for enabling perf build?

Comment (ranj063) [reply app/perf_overlay.conf:1]: Also, Can you please fix your name in the sign-off?

File: src/include/sof/audio/component_ext.h [hunk 5]
```diff
@@ -180,20 +180,12 @@ static inline int comp_copy(struct comp_dev *dev)
 
 	/* copy only if we are the owner of the component */
 	if (cpu_is_me(dev->ipc_config.core)) {
-/* BugLink: https://github.com/zephyrproject-rtos/zephyr/issues/43786
- * TODO: Remove this once the bug gets fixed.
- */
-#ifndef __ZEPHYR__
+#if CONFIG_PERFORMANCE_COUNTERS
```
Comment (lgirdwood) [inline src/include/sof/audio/component_ext.h:183]: Is the above bug fixed by this commit ?

File: src/ipc/ipc-helper.c [hunk 6]
```diff
@@ -40,7 +40,7 @@ struct comp_buffer *buffer_new(const struct sof_ipc_buffer *desc)
 {
 	struct comp_buffer *buffer;
 
-	tr_info(&buffer_tr, "buffer new size 0x%x id %d.%d flags 0x%x",
+	tr_info(&buffer_tr, "perf_mem buffer new size 0x%x id %d.%d flags 0x%x",
```
Comment (lgirdwood) [inline src/ipc/ipc-helper.c:43]: We should put memory based profiling in another patch / PR.

File: app/perf_overlay.conf (hunk 4 above)
Comment (btian1) [reply app/perf_overlay.conf:1]: let me check, I already tried to fix sign off.
ok, seems change too much in one patch is not a good way , let me split based on you, marc, Liam's requirement

**Comment by btian1** [comment]
Thanks all for your review, let me first split it to 2 PRs, will take Rajani and Marc's comments, change perf_cnt as a separate commit.
due to this module will cause the CI failure, it is always there.

File: src/include/sof/audio/component_ext.h (hunk 5 above)
Comment (btian1) [reply src/include/sof/audio/component_ext.h:183]: not yet, let me fix that comp_copy first, then split patch for merge, according to songchao, it caused by xcc compiler, will check the exact error and try to fix it.

**Comment by lgirdwood** [comment]
@singalsu can you work with @btian1 and fold in your recent updates here too. Thanks !

File: src/include/sof/audio/component.h [hunk 7]
```diff
@@ -233,12 +233,14 @@ enum {
 #endif /* #if defined(__ZEPHYR__) && defined(CONFIG_ZEPHYR_LOG) */
 
 #define comp_perf_info(pcd, comp_p)					\
-	comp_info(comp_p, "perf comp_copy peak plat %u cpu %u",		\
+	comp_info(comp_p, "perf_cycle comp_copy peak plat %u cpu %u",		\
```
Comment (kv2019i) [inline src/include/sof/audio/component.h:236]: We don't know all users of these performance traces, so I'd keep the names unchanged unless the semantics change.

File: src/schedule/zephyr_domain.c [hunk 8]
```diff
@@ -84,7 +84,7 @@ static void zephyr_domain_thread_fn(void *p1, void *p2, void *p3)
 
 		if (++runs == 1 << CYCLES_WINDOW_SIZE) {
 			cycles_sum >>= CYCLES_WINDOW_SIZE;
-			tr_info(&ll_tr, "ll timer avg %u, max %u, overruns %u",
+			tr_info(&ll_tr, "perf_cycle ll timer avg %u, max %u, overruns %u",
```
Comment (kv2019i) [inline src/schedule/zephyr_domain.c:87]: I think here I'd prefer to keep old naming unless there's a really good reason to change. We might have downstream users with scripts parsing this data, and this would be causing churn for them.

File: src/ipc/ipc-helper.c (hunk 6 above)
Comment (kv2019i) [reply src/ipc/ipc-helper.c:43]: @btian1 We have heap debug separately and that is probably better place to track the memory usage.

File: src/schedule/zephyr_ll.c [hunk 9]
```diff
@@ -122,30 +122,18 @@ static void zephyr_ll_task_insert_after_unlocked(struct task *task, struct task
 
 static inline enum task_state do_task_run(struct task *task)
 {
-	uint32_t cycles0, cycles1, diff;
 	enum task_state state;
 
-	cycles0 = k_cycle_get_32();
+#if CONFIG_PERFORMANCE_COUNTERS
+	perf_cnt_init(&task->pcd);
+#endif
```
Comment (kv2019i) [inline src/schedule/zephyr_ll.c:129]: @lgirdwood We agreed with @btian1 to get rid of the "always-built-in" traces task level state. I.e. you have to opt-in and build performance counters. You @lgirdwood originally asked for these, so please ack/nack the change.

**Review by kv2019i — COMMENTED** [review]
I'll continue the review in next version. But looks good so far. We need to pay close attention to much trace output we change -- i.e. if there's no need to change formatting of a message, it should be kept. We also need to ensure we are still aligned with these enhancement requests:
 https://github.com/thesofproject/sof/issues/5518 and https://github.com/thesofproject/sof/issues/5521

File: src/schedule/zephyr_domain.c (hunk 8 above)
Comment (singalsu) [reply src/schedule/zephyr_domain.c:87]: Please be consistent with keyword-value pairs. Use everywhere comma as separator or just blanks to help make automatic parsing scripts for trace log. I'd also align the previous perf traces to same format.

File: src/schedule/zephyr_ll.c (hunk 9 above)
Comment (lgirdwood) [reply src/schedule/zephyr_ll.c:129]: ack

Comment (lgirdwood) [reply src/schedule/zephyr_ll.c:129]: Btw, we do need to keep the over budget warnings/errors in the trace for user bug reports.

Comment (kv2019i) [reply src/schedule/zephyr_ll.c:129]: Ack @lgirdwood . We still keep the counter and checks in zephyr_domain_thread_fn(). This emit data on overload load on each DSP core, and emits a warning if the budget is overruns (the "overruns:xx" counter). This has proven useful to debug system issue (e.g. some driver mishaving, unexpected IPC sequences, etc system level issues).

File: src/include/sof/audio/component.h (hunk 7 above)
Comment (btian1) [reply src/include/sof/audio/component.h:236]: ok

File: src/schedule/zephyr_ll.c (hunk 9 above)
Comment (btian1) [reply src/schedule/zephyr_ll.c:129]: ok, will only keep build in for core, for task and component, it will depend on user enable -o app/perf_overlay.conf during build, currently are focusing on the comp_copy from header file to c source file, cmoka have some errors, in fixing.

File: app/perf_overlay.conf (hunk 4 above)
Comment (marc-hb) [reply app/perf_overlay.conf:1]: @andrula-song this overlay is not used by default.

File: xtos/include/sof/lib/perf_cnt.h (hunk 3 above)
Comment (btian1) [reply xtos/include/sof/lib/perf_cnt.h:21]: CONFIG_TIMER_HAS_64BIT_CYCLE_COUNTER this was enabled already, current cpu_ts and plat_ts are all get from:
k_cycle_get_64

32 bit need more implementation to avoid overflow,  it is just for profiling and only be enabled with -o app/perf_overlay.conf.

**Comment by btian1** [comment]
this patch depend on #6344, once #6344 merged, I will update a new patch set to cover above comments

**Comment by lgirdwood** [comment]
@singalsu your script update align with this PR and are all the data points you need available ?

**Comment by singalsu** [comment]
> @singalsu does your script update align with this PR and are all the data points you need available ?

I need to try this myself to confirm it has everything needed. One known limitation is not in this PR but in IPC4 mtrace that does not provide component instance identifiers like sof-logger does. The topologies have multiple gain and mixer parts and we need to know for every instance separately the timer counts.

**Comment by singalsu** [comment]
I get build fail with IPC4 (`./sof/scripts/xtensa-build-zephyr.py -u -s -i IPC4 tgl`):

```
In file included from /home/singalsu/work/current/sof/sof/src/ipc/ipc-helper.c:10:
/home/singalsu/work/current/sof/sof/zephyr/../src/include/sof/audio/component_ext.h: In function ‘comp_copy’:
/home/singalsu/work/current/sof/sof/zephyr/../src/include/sof/audio/component_ext.h:189: error: ‘__log_level’ undeclared (first use in this function)
/home/singalsu/work/current/sof/sof/zephyr/../src/include/sof/audio/component_ext.h:189: error: (Each undeclared identifier is reported only once
/home/singalsu/work/current/sof/sof/zephyr/../src/include/sof/audio/component_ext.h:189: error: for each function it appears in.)
/home/singalsu/work/current/sof/sof/zephyr/../src/include/sof/audio/component_ext.h:189: error: ‘__log_current_const_data’ undeclared (first use in this function)
/home/singalsu/work/current/sof/sof/zephyr/../src/include/sof/audio/component_ext.h:189: error: ‘__log_current_dynamic_data’ undeclared (first use in this function)
```

File: src/include/sof/audio/component.h [hunk 10]
```diff
@@ -233,12 +233,14 @@ enum {
...
 		  (uint32_t)((pcd)->plat_delta_peak),			\
 		  (uint32_t)((pcd)->cpu_delta_peak))
 
 #define comp_perf_avg_info(pcd, comp_p)					\
-	comp_info(comp_p, "perf comp_copy cpu avg %u (current peak %u)",\
+	comp_info(comp_p, "perf_cycle comp_copy samples %u period %u cpu avg %u  peak %u",\
```
Comment (singalsu) [inline src/include/sof/audio/component.h:241]: There's one extra blank in " peak". It's nicer for automatic parsing if there's no need to remove multiple successive blanks. It's good that you use only blanks as separators and removed the above () characters.

**Comment by singalsu** [comment]
> @singalsu does your script update align with this PR and are all the data points you need available ?

This patch works OK with my script with IPC3 build. Here's an example of a playback to PCM0P device with sof-hda-generic.tplg.

Screenshot from 2022-10-07 16-15-39

File: app/perf_overlay.conf (hunk 4 above)
Comment (singalsu) [inline app/perf_overlay.conf:1]: Nit, start commit message sentences with capital letters. Doesn't checkpatch complain about this.

**Review by singalsu — CHANGES_REQUESTED** [review]

File: src/include/sof/audio/component.h [hunk 11]
```diff
@@ -238,7 +238,9 @@ enum {
 		  (uint32_t)((pcd)->cpu_delta_peak))
 
 #define comp_perf_avg_info(pcd, comp_p)					\
-	comp_info(comp_p, "perf comp_copy cpu avg %u (current peak %u)",\
+	comp_info(comp_p, "perf comp_copy samples %u period %u cpu avg %u peak %u",\
+		  (uint32_t)((comp_p)->frames),            \
+		  (uint32_t)((comp_p)->period),			    \
```
Comment (marc-hb) [inline src/include/sof/audio/component.h:243]: `frames` and `period` are defined as `uint32_t` in `src/include/sof/audio/component.h`. Why are they cast to `uint32_t`? Casts hide bugs from the compiler, we want fewer casts not more.

Comment (marc-hb) [reply src/include/sof/audio/component.h:243]: BTW renaming `comp_p` to `comp_dev_p` would save a lot of time to people new to this code (which involves an unreasonable number of macro indirections for "functional programming in C")

**Review by marc-hb — CHANGES_REQUESTED** [review]

File: app/perf_overlay.conf (hunk 4 above)
Comment (kv2019i) [reply app/perf_overlay.conf:1]: @btian1 The signed-off is still wrong, you should have proper full name there. 
Also, I recommend @ranj063 's idea of splitting to two patches. One to move the task perf tracking to new interface and a separate patch to add the performance build.

**Review by kv2019i — CHANGES_REQUESTED** [review]
Still some unresolved issues in review, please see inline.

**Comment by btian1** [comment]
due to this patch cover too much, I closed this PR and submit 3 new PR for your review, please goto:
#6406 
#6407 
#6408 

to continue your review, this PR is closing.

Thanks
Tim
